crates/storage-pg/src/queue/worker.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2024 New Vector Ltd.
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+// Please see LICENSE in the repository root for full details.
+
+//! A module containing the PostgreSQL implementation of the
+//! [`QueueWorkerRepository`].
+
+use async_trait::async_trait;
+use chrono::Duration;
+use mas_storage::{
+    queue::{QueueWorkerRepository, Worker},
+    Clock,
+};
+use rand::RngCore;
+use sqlx::PgConnection;
+use ulid::Ulid;
+use uuid::Uuid;
+
+use crate::{DatabaseError, ExecuteExt};
+
+/// An implementation of [`QueueWorkerRepository`] for a PostgreSQL connection.
+pub struct PgQueueWorkerRepository<'c> {
+    conn: &'c mut PgConnection,
+}
+
+impl<'c> PgQueueWorkerRepository<'c> {
+    /// Create a new [`PgQueueWorkerRepository`] from an active PostgreSQL
+    /// connection.
+    #[must_use]
+    pub fn new(conn: &'c mut PgConnection) -> Self {
+        Self { conn }
+    }
+}
+
+#[async_trait]
+impl<'c> QueueWorkerRepository for PgQueueWorkerRepository<'c> {
+    type Error = DatabaseError;
+
+    #[tracing::instrument(
+        name = "db.queue_worker.register",
+        skip_all,
+        fields(
+            worker.id,
+            db.query.text,
+        ),
+        err,
+    )]
+    async fn register(
+        &mut self,
+        rng: &mut (dyn RngCore + Send),
+        clock: &dyn Clock,
+    ) -> Result<Worker, Self::Error> {
+        let now = clock.now();
+        let worker_id = Ulid::from_datetime_with_source(now.into(), rng);
+        tracing::Span::current().record("worker.id", tracing::field::display(worker_id));
+
+        sqlx::query!(
+            r#"
+                INSERT INTO queue_workers (queue_worker_id, registered_at, last_seen_at)
+                VALUES ($1, $2, $2)
+            "#,
+            Uuid::from(worker_id),
+            now,
+        )
+        .traced()
+        .execute(&mut *self.conn)
+        .await?;
+
+        Ok(Worker { id: worker_id })
+    }
+
+    #[tracing::instrument(
+        name = "db.queue_worker.heartbeat",
+        skip_all,
+        fields(
+            %worker.id,
+            db.query.text,
+        ),
+        err,
+    )]
+    async fn heartbeat(
+        &mut self,
+        clock: &dyn Clock,
+        worker: Worker,
+    ) -> Result<Worker, Self::Error> {
+        let now = clock.now();
+        let res = sqlx::query!(
+            r#"
+                UPDATE queue_workers
+                SET last_seen_at = $2
+                WHERE queue_worker_id = $1 AND shutdown_at IS NULL
+            "#,
+            Uuid::from(worker.id),
+            now,
+        )
+        .traced()
+        .execute(&mut *self.conn)
+        .await?;
+
+        // If no row was updated, the worker was shutdown so we return an error
+        DatabaseError::ensure_affected_rows(&res, 1)?;
+
+        Ok(worker)
```

Comment:
the docstring says this returns the modified worker, but I don't see us modifying it.

I would expect the worker to track its own validity timestamps, but I guess the critical thing here is just that we 'take away' the Worker if we can't renew it?